builtins.go

```diff
@@ -1194,6 +1195,34 @@ func builtinParseJSON(i *interpreter, str value) (value, error) {
 	return jsonToValue(i, parsedJSON)
 }
 
+func builtinParseYAML(i *interpreter, str value) (value, error) {
+	sval, err := i.getString(str)
+	if err != nil {
+		return nil, err
+	}
+	s := sval.getGoString()
+
+	isYamlStream := strings.Contains(s, "---")
```

Comment:
The shotgun parsing is a bit gross, I agree. Anyone adding those characters to a YAML document will be in for a difficult time if they aren't escaping it thoroughly. The `.Contains` check can probably be made more explicit to check for start of line or end of line characters if that makes you feel more comfortable?

Alternatively, we can adjust the user facing API. The only reason for "sniffing" whether a document is to be perhaps more user-friendly by converting a single YAML document to an object `{}` and a stream of YAML documents to an array `[]`. An alternative user API could be to require a user to choose. e.g. `parseYaml` that returns a single object `{}` and `parseYamlStream` which returns an array.

Reply:
I see. I think you could achieve the same effect in a cleaner way just by checking `len(elems)`. The elems will only have multiple elements when there are multiple documents. If it's an array, it will still be just one document. Is that correct?

> An alternative user API could be to require a user to choose. e.g. parseYaml that returns a single object {} and parseYamlStream which returns an array.

That could work. It's more complicated, but cleaner in a way. I don't have a strong opinion. We can also add parseYamlStream later (which forces returning of an array, even if there is only one document).

Reply:
> I see. I think you could achieve the same effect in a cleaner way just by checking `len(elems)`.

How do we differentiate between a single YAML document with a scalar list vs a YAML stream of documents?

YAML, like JSON, can have a top level sequence or array. Some examples.

```
// Single doc, scalar array at root
std.parseYaml(
|||
  - {a: 1, b: 2}
  - {a: 3, b: 4}
  - {a: 5, b: 6}
|||
)

[
   {
      "a": 1,
      "b": 2
   },
   {
      "a": 3,
      "b": 4
   },
   {
      "a": 5,
      "b": 6
   }
]
```

```
// Mutli doc YAML stream with document start separators.
std.parseYaml(
  |||
   ---
   {a: 1, b: 2}
   ---
   {a: 3, b: 4}
   ---
   {a: 5, b: 6}
  |||
)

[
   {
      "a": 1,
      "b": 2
   },
   {
      "a": 3,
      "b": 4
   },
   {
      "a": 5,
      "b": 6
   }
]
```

```
// Single doc, scalar array at root. Indentation instead of {
std.parseYaml(
|||
  - a: 1
    b: 2
  - a: 3
    b: 4
  - a: 5
    b: 6
|||
)

[
   {
      "a": 1,
      "b": 2
   },
   {
      "a": 3,
      "b": 4
   },
   {
      "a": 5,
      "b": 6
   }
]
```

Since there can be scalar arrays at the root, checking the `len(elems)` isn't enough because it isn't possible to tell the difference between a yaml stream with 1 document `len(elems) == 1` and a single document with a top-level scalar array of length 1.

The approach used in the current implementation assumes that if you specify a YAML stream using `---` you know you will receive an array of 1 or more documents. If you do not specify a YAML stream, you will get whatever YAML object you specified, either a Map or a Seq of length, 0, 1 or more elements.

YAML is complicated! :)

Reply:
> Since there can be scalar arrays at the root, checking the len(elems) isn't enough because it isn't possible to tell the difference between a yaml stream with 1 document len(elems) == 1 and a single document with a top-level scalar array of length 1.

Please correct me if I misunderstand something. I assumed that `d.Decode(&elem)` reads one document. So len(elems) will be the number of documents in a stream. If it's only one document (even if it's an array) it would read the whole thing. Is that correct? 

Is your point that:
```
---
foo: bar
```
and
```
foo: bar
```
both produce `{foo: "bar"}`, but the user might have expected the array in the first case? Hnnnn... I think it's  a valid concern, because when processing streams, it makes it hard to handle special case of 1-element stream. 

Sniffing for `---` still seems pretty bad (even if we made sure that it's a line on its own). First, I'm afraid of special cases. Second, IIUC formally it's not a magic symbol for streams, but an explicit start of the document.

Perhaps an explicit `parseYamlStream` which always returns the array is the way out of this.

@sparkprime @sh0rez Any thoughts about handling YAML streams?

Reply:
> both produce `{foo: "bar"}`, but the user might have expected the array in the first case? Hnnnn... I think it's a valid concern, because when processing streams, it makes it hard to handle special case of 1-element stream.

Yes, that's my concern. You are correct in that it attempts to read a single document, but should we treat a 1-element stream differently or not is the question? Should the output of the following 2 examples be different or the same?

```
std.parseYaml(
  |||
   ---
   foo: bar
  |||
)
```

```
std.parseYaml(
|||
   foo: bar
|||
)
```

What about this one?

```
std.parseYaml(
  |||
   ---
   foo: bar
   ---
   wibble: wobble
  |||
)
```

```
std.parseYaml(
|||
  - foo: bar
  - wibble: wobble
|||
)
```

I think I can be convinced either way to be honest.